File: src/main/java/com/ebuozturk/ecommerce/model/Category.java
package com.ebuozturk.ecommerce.model;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;


@Entity
@Getter
@EqualsAndHashCode
@NoArgsConstructor
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;

    @OneToMany(mappedBy = "category")
    private final Set<Product> products = new HashSet<>();

    public Category(String name) {
        this.name = name;
    }

    public Category(Long id, String name) {
        this.id = id;
        this.name = name;
    }
}

File: src/main/java/com/ebuozturk/ecommerce/model/Category.kt
package com.ebuozturk.ecommerce.model;

import javax.persistence.*;

@Entity
data class Category @JvmOverloads constructor(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id:Long? = null,
    val name:String,
    @OneToMany(fetch = FetchType.LAZY,mappedBy = "category")
    val products: Set<Product>
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Category

        if (id != null && id != other.id) return false
        if (name != other.name) return false
        if (products != other.products) return false

        return true
    }

}
